File: app/src/main/java/wxm/KeepAccount/utility/PreferencesUtil.java
package wxm.KeepAccount.utility;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;

import wxm.KeepAccount.R;
import wxm.androidutil.util.UiUtil;

/**
 * preference helper
 * Created by 123 on 2016/6/18.
 */
public class PreferencesUtil {
    public final static String SET_PAY_COLOR = "pay";
    public final static String SET_INCOME_COLOR = "income";
    public final static String SET_BUDGET_UESED_COLOR = "budget_used";
    public final static String SET_BUDGET_BALANCE_COLOR = "budget_balance";
    private final static String PROPERTIES_NAME = "keepaccount_properties";
    private final static String SET_HOT_ACTION = "hot_action";
    private final static String SET_CHART_COLOR = "chart_color";

    /// BEGIN

    /**
     * load first page action setting
     * @return      action in first page
     */
    public static List<String> loadHotAction() {
        SharedPreferences param = ContextUtil.getInstance()
                .getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE);

        String def = String.format(Locale.CHINA,
                "%s:%s:%s:%s:%s:%s",
                ActionHelper.ACT_ADD_DATA, ActionHelper.ACT_LOOK_DATA,
                ActionHelper.ACT_CALENDAR_VIEW, ActionHelper.ACT_LOOK_BUDGET,
                ActionHelper.ACT_ADD_BUDGET, ActionHelper.ACT_LOGOUT);

        String load = param.getString(SET_HOT_ACTION, def);
        return parsePreferences(load);
    }


    /**
     * save fist page action
     * @param acts  action in fist page
     */
    public static void saveHotAction(List<String> acts) {
        SharedPreferences param = ContextUtil.getInstance()
                .getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE);

        String pr = parseToPreferences(acts);
        param.edit().putString(SET_HOT_ACTION, pr).apply();
    }
    /// END


    /// BEGIN

    /**
     * load chart color setting
     * @return  color setting
     */
    public static HashMap<String, Integer> loadChartColor() {
        SharedPreferences param = ContextUtil.getInstance()
                .getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE);

        Context ct = ContextUtil.getInstance();
        String sb = SET_PAY_COLOR + ":" + String.valueOf(UiUtil.getColor(ct, R.color.sienna)) +
                " " + SET_INCOME_COLOR + ":" + String.valueOf(UiUtil.getColor(ct, R.color.teal)) +
                " " + SET_BUDGET_UESED_COLOR + ":" + String.valueOf(UiUtil.getColor(ct, R.color.sienna)) +
                " " + SET_BUDGET_BALANCE_COLOR + ":" + String.valueOf(UiUtil.getColor(ct, R.color.teal));

        String load = param.getString(SET_CHART_COLOR, sb);
        return parseChartColors(load);
    }


    /**
     * save chart color setting
     * @param ccs   color setting
     */
    public static void saveChartColor(HashMap<String, Integer> ccs) {
        SharedPreferences param = ContextUtil.getInstance()
                .getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE);

        String pr = parseChartColorsToString(ccs);
        param.edit().putString(SET_CHART_COLOR, pr).apply();
    }
    /// END


    /// BEGIN PRIVATE

    /**
     * parse setting string
     * @param pr    setting string
     * @return      settings
     */
    private static List<String> parsePreferences(String pr) {
        ArrayList<String> ret = new ArrayList<>();
        String[] pr_arr = pr.split(":");
        Collections.addAll(ret, pr_arr);

        return ret;
    }

    /**
     * parse settings to string
     * @param acts      settings
     * @return          setting string
     */
    private static String parseToPreferences(List<String> acts) {
        boolean ff = true;
        StringBuilder sb = new StringBuilder();
        for (String i : acts) {
            if (!ff) {
                sb.append(":");
            } else {
                ff = false;
            }

            sb.append(i);
        }

        return sb.toString();
    }

    /**
     * parse setting string for chart color
     * @param cc    setting string
     * @return      chart color setting
     */
    private static HashMap<String, Integer> parseChartColors(String cc) {
        HashMap<String, Integer> ret = new HashMap<>();
        String[] cclns = cc.split(" ");
        for (String i : cclns) {
            String[] iilns = i.split(":");

            ret.put(iilns[0], Integer.parseInt(iilns[1]));
        }

        return ret;
    }


    /**
     * parse chart color to setting string
     * @param hmcc      chart color setting
     * @return          setting string
     */
    private static String parseChartColorsToString(HashMap<String, Integer> hmcc) {
        StringBuilder ret = new StringBuilder();
        for (String i : hmcc.keySet()) {
            StringBuilder sb = new StringBuilder();
            sb.append(i).append(":").append(hmcc.get(i).toString());

            if (0 == ret.length())
                ret.append(sb);
            else
                ret.append(" ").append(sb);
        }

        return ret.toString();
    }
    /// END PRIVATE
}

File: app/src/main/java/wxm/KeepAccount/utility/PreferencesUtil.kt
package wxm.KeepAccount.utility

import android.content.Context
import wxm.KeepAccount.R
import wxm.KeepAccount.define.EAction
import java.util.*

/**
 * preference helper
 * Created by WangXM on 2016/6/18.
 */
object PreferencesUtil {
    const val SET_PAY_COLOR = "pay"
    const val SET_INCOME_COLOR = "income"
    const val SET_BUDGET_UESED_COLOR = "budget_used"
    const val SET_BUDGET_BALANCE_COLOR = "budget_balance"
    private const val PROPERTIES_NAME = "keepaccount_properties"
    private const val SET_HOT_ACTION = "hot_action"
    private const val SET_CHART_COLOR = "chart_color"

    /// BEGIN
    /**
     * load first page action setting
     * @return      action in first page
     */
    fun loadHotAction(): List<String> {
        val param = AppUtil.self.getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE)
        return String.format(Locale.CHINA,
                "%s:%s:%s:%s:%s:%s",
                EAction.ADD_DATA.actName, EAction.LOOK_DATA.actName,
                EAction.CALENDAR_VIEW.actName, EAction.LOOK_BUDGET.actName,
                EAction.ADD_BUDGET.actName, EAction.LOGOUT.actName).let {
            param.getString(SET_HOT_ACTION, it)
        }.let {
            parsePreferences(it)
        }
    }


    /**
     * save fist page action
     * @param acts  action in fist page
     */
    fun saveHotAction(acts: List<String>) {
        AppUtil.self.getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE)
                .edit().putString(SET_HOT_ACTION, parseToPreferences(acts)).apply()
    }
    /// END


    /// BEGIN
    /**
     * load chart color setting
     * @return  color setting
     */
    fun loadChartColor(): HashMap<String, Int> {
        return AppUtil.self.let {
            "$SET_PAY_COLOR:${it.getColor(R.color.sienna)}" +
                    " $SET_INCOME_COLOR:${it.getColor(R.color.teal)}" +
                    " $SET_BUDGET_UESED_COLOR:${it.getColor(R.color.sienna)}" +
                    " $SET_BUDGET_BALANCE_COLOR:${it.getColor(R.color.teal)}"
        }.let {
            AppUtil.self
                    .getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE)
                    .getString(SET_CHART_COLOR, it)!!
        }.let {
            parseChartColors(it)
        }
    }


    /**
     * save chart color setting
     * @param ccs   color setting
     */
    fun saveChartColor(ccs: HashMap<String, Int>) {
        AppUtil.self.getSharedPreferences(PROPERTIES_NAME, Context.MODE_PRIVATE).apply {
            edit().putString(SET_CHART_COLOR, parseChartColorsToString(ccs)).apply()
        }
    }
    /// END


    /// BEGIN PRIVATE
    /**
     * use [delimiter] split lns to arry
     */
    private fun splitString(lns: String, delimiter: String): Array<String> {
        return lns.split(delimiter.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
    }

    /**
     * parse setting string
     * @param pr    setting string
     * @return      settings
     */
    private fun parsePreferences(pr: String): List<String> {
        return ArrayList<String>().apply {
            Collections.addAll(this, *splitString(pr, ":"))
        }
    }

    /**
     * parse settings to string
     * @param acts      settings
     * @return          setting string
     */
    private fun parseToPreferences(acts: List<String>): String {
        var ff = true
        val sb = StringBuilder()
        acts.forEach {
            if (!ff) {
                sb.append(":")
            } else {
                ff = false
            }

            sb.append(it)
        }

        return sb.toString()
    }

    /**
     * parse setting string for chart color
     * @param cc    setting string
     * @return      chart color setting
     */
    private fun parseChartColors(cc: String): HashMap<String, Int> {
        return HashMap<String, Int>().apply {
            splitString(cc, " ").forEach {
                splitString(it, ":").let {
                    put(it[0], Integer.parseInt(it[1]))
                }
            }
        }
    }

    /**
     * parse chart color to setting string
     * @param hmCC      chart color setting
     * @return          setting string
     */
    private fun parseChartColorsToString(hmCC: HashMap<String, Int>): String {
        val ret = StringBuilder()
        hmCC.forEach { t, u ->
            StringBuilder().apply {
                append(t).append(":").append(u.toString())
            }.let {
                if (ret.isEmpty())
                    ret.append(it)
                else
                    ret.append(" ").append(it)

                Unit
            }
        }

        return ret.toString()
    }
    /// END PRIVATE
}
